core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala

```diff
@@ -0,0 +1,297 @@
+    TestUtils.waitUntilTrue(() => !zkUtils.pathExists(ZkUtils.ReassignPartitionsPath),
+      "failed to remove reassign partitions path after completion")
+  }
+
+  @Test
+  def testPreferredReplicaLeaderElection(): Unit = {
+    servers = makeServers(2)
+    val tp = TopicAndPartition("t", 0)
+    val assignment = Map(tp.partition -> Seq(1, 0))
+    TestUtils.createTopic(zkUtils, tp.topic, partitionReplicaAssignment = assignment, servers = servers)
+    servers(1).shutdown()
+    servers(1).awaitShutdown()
+    waitForPartitionState(tp, KafkaController.InitialControllerEpoch, 0, LeaderAndIsr.initialLeaderEpoch + 1,
+      "failed to get expected partition state upon broker shutdown")
+    servers(1).startup()
+    TestUtils.waitUntilTrue(() => servers.forall(_.metadataCache.isBrokerAlive(1)), "broker join was not broadcasted to the cluster")
+    zkUtils.createPersistentPath(ZkUtils.PreferredReplicaLeaderElectionPath, ZkUtils.preferredReplicaLeaderElectionZkData(Set(tp)))
+    TestUtils.waitUntilTrue(() => !zkUtils.pathExists(ZkUtils.PreferredReplicaLeaderElectionPath),
+      "failed to remove preferred replica leader election path after completion")
+    waitForPartitionState(tp, KafkaController.InitialControllerEpoch, 1, LeaderAndIsr.initialLeaderEpoch + 2,
+      "failed to get expected partition state upon broker startup")
+  }
+
+  @Test
+  def testPreferredReplicaLeaderElectionWithOfflinePreferredReplica(): Unit = {
+    servers = makeServers(2)
+    val tp = TopicAndPartition("t", 0)
+    val assignment = Map(tp.partition -> Seq(1, 0))
+    TestUtils.createTopic(zkUtils, tp.topic, partitionReplicaAssignment = assignment, servers = servers)
+    servers(1).shutdown()
+    servers(1).awaitShutdown()
+    zkUtils.createPersistentPath(ZkUtils.PreferredReplicaLeaderElectionPath, ZkUtils.preferredReplicaLeaderElectionZkData(Set(tp)))
+    TestUtils.waitUntilTrue(() => !zkUtils.pathExists(ZkUtils.PreferredReplicaLeaderElectionPath),
+      "failed to remove preferred replica leader election path after giving up")
+    waitForPartitionState(tp, KafkaController.InitialControllerEpoch, 0, LeaderAndIsr.initialLeaderEpoch + 1,
+      "failed to get expected partition state upon broker shutdown")
+  }
+
+  @Test
+  def testAutoPreferredReplicaLeaderElection(): Unit = {
+    servers = makeServers(2, autoLeaderRebalanceEnable = true)
+    val tp = TopicAndPartition("t", 0)
+    val assignment = Map(tp.partition -> Seq(1, 0))
+    TestUtils.createTopic(zkUtils, tp.topic, partitionReplicaAssignment = assignment, servers = servers)
+    servers(1).shutdown()
+    servers(1).awaitShutdown()
+    waitForPartitionState(tp, KafkaController.InitialControllerEpoch, 0, LeaderAndIsr.initialLeaderEpoch + 1,
+      "failed to get expected partition state upon broker shutdown")
+    servers(1).startup()
+    waitForPartitionState(tp, KafkaController.InitialControllerEpoch, 1, LeaderAndIsr.initialLeaderEpoch + 2,
```

Comment:
Hmm, by default, auto leader balancing will only kick in once every 5 mins. We probably want to reduce leader.imbalance.check.interval.seconds and leader.imbalance.per.broker.percentage in the test.